src/main/java/io/cdap/plugin/servicenow/source/ServiceNowBaseSourceConfig.java

```diff
@@ -0,0 +1,327 @@
+      return;
+    }
+
+    if (Util.isNullOrEmpty(clientId)) {
+      collector.addFailure("Client ID must be specified.", null)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_CLIENT_ID);
+    }
+
+    if (Util.isNullOrEmpty(clientSecret)) {
+      collector.addFailure("Client Secret must be specified.", null)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_CLIENT_SECRET);
+    }
+
+    if (Util.isNullOrEmpty(restApiEndpoint)) {
+      collector.addFailure("API Endpoint must be specified.", null)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_API_ENDPOINT);
+    }
+
+    if (Util.isNullOrEmpty(user)) {
+      collector.addFailure("User name must be specified.", null)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_USER);
+    }
+
+    if (Util.isNullOrEmpty(password)) {
+      collector.addFailure("Password must be specified.", null)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_PASSWORD);
+    }
+
+    validateServiceNowConnection(collector);
+  }
+
+
+  @VisibleForTesting
+  void validateServiceNowConnection(FailureCollector collector) {
+    try {
+      ServiceNowTableAPIClientImpl restApi = new ServiceNowTableAPIClientImpl(this);
+      restApi.getAccessToken();
+    } catch (Exception e) {
+      collector.addFailure("Unable to connect to ServiceNow Instance.",
+          "Ensure properties like Client ID, Client Secret, API Endpoint, User Name, Password " +
+            "are correct.")
+        .withConfigProperty(ServiceNowConstants.PROPERTY_CLIENT_ID)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_CLIENT_SECRET)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_API_ENDPOINT)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_USER)
+        .withConfigProperty(ServiceNowConstants.PROPERTY_PASSWORD)
+        .withStacktrace(e.getStackTrace());
+    }
+  }
+
+  /**
+   * Returns the value type chosen.
+   *
+   * @param collector The failure collector to collect the errors
+   * @return An instance of SourceValueType
+   */
+  public SourceValueType getValueType(FailureCollector collector) {
+    SourceValueType type = getValueType();
+    if (type != null) {
+      return type;
+    }
+
+    collector.addFailure("Unsupported type value: " + valueType,
+        String.format("Supported value types are: %s", SourceValueType.getSupportedValueTypes()))
+      .withConfigProperty(ServiceNowConstants.PROPERTY_VALUE_TYPE);
+    collector.getOrThrowException();
+    return null;
+  }
+
+  /**
+   * Returns the value type chosen.
+   *
+   * @return An instance of SourceValueType
+   */
+  @Nullable
+  public SourceValueType getValueType() {
+    Optional<SourceValueType> sourceValueType = SourceValueType.fromValue(valueType);
+
+    return sourceValueType.isPresent() ? sourceValueType.get() : null;
+  }
+
+  public void validateValueType(FailureCollector collector) {
+    if (containsMacro(ServiceNowConstants.PROPERTY_VALUE_TYPE)) {
+      return;
+    }
+
```

Comment:
looks like you aren't really doing any validation here

Reply:
A macro check is also a form of validation, right? Further there is null check in the method `getValueType(FailureCollector collector)` called from within the method `validateValueType(FailureCollector collector)`